imap/itip_support.c

```diff
@@ -403,36 +403,56 @@ HIDDEN icalcomponent *master_to_recurrence(icalcomponent *master,
 }
 
 
-static const char *deliver_merge_reply(icalcomponent *ical,
-                                       icalcomponent *reply)
+static const char *deliver_merge_reply(icalcomponent *ical,  // current iCalendar
+                                       icalcomponent *reply) // iTIP reply
 {
-    struct hash_table comp_table;
+    struct hash_table override_table;
     icalcomponent *comp, *itip, *master = NULL;
     icalcomponent_kind kind;
     icalproperty *prop, *att;
     icalparameter *param;
     icalparameter_partstat partstat = ICAL_PARTSTAT_NONE;
     const char *attendee = NULL, *cn = NULL;
-    icaltimezone *utc_zone = icaltimezone_get_utc_timezone();
     icaltimezone *startzone = NULL;
     icaltimetype dtstart;
+    struct icalrecurrencetype rrule = ICALRECURRENCETYPE_INITIALIZER;
+    struct hash_table rdate_table = HASH_TABLE_INITIALIZER;
 
-    /* Add each component of old object to hash table for comparison */
-    construct_hash_table(&comp_table, 10, 1);
+    /* Add each override component of current object to hash table for lookup */
+    construct_hash_table(&override_table, 10, 1);
     comp = icalcomponent_get_first_real_component(ical);
     kind = icalcomponent_isa(comp);
     do {
         icaltimetype recurid = icalcomponent_get_recurrenceid_with_zone(comp);
 
-        if (icaltime_is_null_time(recurid)) {
+        if (!icaltime_is_null_time(recurid)) {
+            hash_insert(icaltime_as_ical_string(recurid), comp, &override_table);
+        }
+        else {
+            /* Master component - get recurrence info */
             master = comp;
             dtstart = icalcomponent_get_dtstart(master);
             startzone = (icaltimezone *) icaltime_get_timezone(dtstart);
-        }
-        else {
-            /* Convert RECURRENCE-ID to UTC */
-            recurid = icaltime_convert_to_zone(recurid, utc_zone);
-            hash_insert(icaltime_as_ical_string(recurid), comp, &comp_table);
+
+            /* Get RRULE, if any */
+            prop = icalcomponent_get_first_property(master, ICAL_RRULE_PROPERTY);
+            if (prop) rrule = icalproperty_get_rrule(prop);
```

Comment:
This will work for 99.9% of all events, but it does not handle multiple RRULEs. We already have support for multiple recurrence rules:

- ical_support.c defines `struct multirrule_iterator` which we would need to export
- `icalcomponent_myforeach` uses the multirrule, but you would need to strip the RDATEs first from the iCalendar data

Probably for the current patch, it would just be enough to gather all rrule iterators into a list and check each once, as you do not need to generate multiple values from a rrule anyway.

Reply:
Done.  Also uses just single loop to fetch all recurrence properties